core: stoppable provisioners, helper/schema for provisioners

mitchellh opened the pull request:
⚠️  For **Terraform 0.9** (Do not merge until master is 0.9) ⚠️ 

**Fast provisioner cancellation!** 💃 

![asdf](https://cloud.githubusercontent.com/assets/1299/21511753/93b67712-cc59-11e6-9b9a-b2ff55df07c4.gif)

This PR includes two things:

  * Create a new Stop API for provisioners that is called when terraform's Stop API (triggered by things like interrupts) is called.

  * Introduce a `helper/schema` framework for provisioners. This makes it easy to consume the Stop API also introduced in this PR. 

This PR uses `context` heavily to make it easy to consume the cancellation, and helper/schema handles managing this context. Terraform's core stop mechanism was also changed to use `context`.

jbardin commented on helper/schema/schema.go:
```diff
@@ -118,9 +118,16 @@ type Schema struct {
+	// and promote them to a list. For example "foo" would be promoted to
+	// ["foo"] automatically. This is primarily for legacy reasons and the
+	// ambiguity is not recommended for new usage. Promotion is only allowed
+	// for primitive element types.
```
Interesting idea, but I don't understand why we run into this. Did we allow defining single elements for a list in the config previously?

VladRassokhin replied:
That's the case of `inline` parameter of 'remote-exec' provisioner, see what is removed in `builtin/provisioners/remote-exec/resource_provisioner.go`

mitchellh replied:
Yeah, backwards compatibility here is the reason this was necessary.

jbardin commented:
sorry, didn't mean to approve yet. still reviewing some, so please don't merge ...

jbardin commented on terraform/context.go:
```diff
@@ -634,30 +633,34 @@ func (c *Context) Refresh() (*State, error) {
 func (c *Context) Stop() {
+	log.Printf("[WARN] terraform: Stop called, initiating interrupt sequence")
+
 	c.l.Lock()
```
there's no special-case lock handling here, so Unlock should be defered.

mitchellh replied:
Done!

jbardin commented on terraform/context.go:
```diff
@@ -634,30 +633,34 @@ func (c *Context) Refresh() (*State, error) {
-	c.sh.Stop()
+		// Stop the context
+		c.runContextCancel()
+		c.runContextCancel = nil
```
Do we need to set this to nil? This one is OK right now I think, but sets the code of for races in the future,

mitchellh replied:
We have to to avoid double-cancels. I'm not sure if that actually has an effect on context though. See Stop()

jbardin commented on terraform/context.go:
```diff
@@ -892,7 +906,15 @@ func (c *Context) walk(
+	// If this is called during a proper run operation, this will never
+	// be nil.
+	var stopCh <-chan struct{}
+	if ctx := c.runContext; ctx != nil {
```
This definitely races with `acquireRun`, which sets the context fields under a mutex. What we can do here is have `walk` get the `stopCh` and pass it in to avoid reading that field without the mutex:

    func (c *Context) watchStop(walker *ContextGraphWalker, stopCh, doneCh <-chan struct{})

jbardin replied:
I'm not a fan of special casing this during tests either; can we ensure that it's set for tests too and get rid of these nil checks?

mitchellh replied:
Its safe to access runContext because an entire Run is wrapped in a lock (acquireRun acquires a lock via cond vars). Fixing the special case though!

jbardin replied:
yes, but `watchStop` doesn't take a lock, and can race with any method writing to the context fields. Just verified this shows up readily with the race detector.

This should fix it by removing the read of that field from watchStop entirely:

```
@@ -811,7 +811,9 @@ func (c *Context) walk(

        // Watch for a stop so we can call the provider Stop() API.
        doneCh := make(chan struct{})
-       go c.watchStop(walker, doneCh)
+       stopCh := c.runContext.Done()
+
+       go c.watchStop(walker, stopCh, doneCh)

        // Walk the real graph, this will block until it completes
        realErr := graph.Walk(walker)
@@ -906,12 +908,7 @@ func (c *Context) walk(
        return walker, realErr
 }

-func (c *Context) watchStop(walker *ContextGraphWalker, doneCh <-chan struct{}) {
-       // Get the stop channel. runContext will never be nil since this should
-       // only be called within the context of an operation started with
-       // acquireRun
-       stopCh := c.runContext.Done()
-
+func (c *Context) watchStop(walker *ContextGraphWalker, stopCh, doneCh <-chan struct{}) {
        // Wait for a stop or completion
        select {
        case <-stopCh:
```

mitchellh commented:
Addressed feedback. Can you check the latest commits? 

All the `run*` unexported vars are safe to access without a lock since `acquireRun`/`releaseRun` protect access. Even if `Stop` is called, another operation can't start until `releaseRun` completes which requires that watchStop/walk and friends aren't running anymore.

jbardin commented on builtin/provisioners/local-exec/resource_provisioner.go:
```diff
@@ -62,8 +68,24 @@ func (p *ResourceProvisioner) Apply(
+		case <-doneCh:
+		case <-ctx.Done():
+			cmd.Process.Kill()
+			err = cmd.Wait()
```
this `err` assignment races with the one in the goroutine above

mitchellh commented:
All current feedback addressed. Please review again. :)
